### PR TITLE
[LT] Integrating with DDP with c10d comm ops falling back to eager

### PR DESCRIPTION
**Pull request by alanwaketan**
This pull request/branch elaborates a prototype that makes all c10d communicative ops fallback to eager to support DDP with LazyTensor. This is a very naive prototype to demonstrate a way how DDP can work with LazyTensor.

Pros:
1. All allReduce ops would be executed in the exact same order as the bucketing algorithm in DDP schedule.

Cons:
1. It forces to break the graph into small pieces whenever there is an allReduce. So theoretically it would have worse performance than the approach where we can capture the full graph. The full graph approach will be the next prototype.
2. For materializing the tensors, it chooses to use the `.to('cuda:i')` method which in the current design of LazyTensor will actually copy the tensor even the tensor is already in the correct cuda device.
3. It might not overlap the communication with the computation well given the allReduces are probably triggered during the trace time instead of the execution time. Need more evidences to prove this theory.

Test Plan:
`LTC_TS_CUDA=1 python lazy_tensor_core/test/ddp/ddp.py`

### CODE REVIEW

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 1]
```diff
@@ -1234,7 +1234,7 @@ void check_gpu_tensors(const std::vector<at::Tensor>& tensors) {
       TORCH_CHECK(false, "Tensors must have identical strides");
     }
     if (!t.is_non_overlapping_and_dense()) {
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1236]: do you have an idea about what to change this check to in the end-state, or how to make it supported for lazy?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 2]
```diff
@@ -1196,7 +1196,7 @@ namespace {
 // Check validity of tensor
 void check_gpu_single_tensor(const at::Tensor& tensor) {
   if (!tensor.is_cuda() || tensor.is_sparse()) {
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1198]: checks like this should be pretty straightforward to just add cases for .is_lazy() || .is_xla()

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp [hunk 3]
```diff
@@ -1338,6 +1338,13 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupNCCL::collective(
     OpType opType,
     const char* profilingTitle) {
 
+  // A hack to copy lazy to cuda to bypass following checks.
+  if (inputs.front().device().type() == at::kLazy) {
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1342]: so is this the part you mentioned about falling back before the 'real' part of DDP happens?

you first do all of the compute on lazy device, and then here you materialize all the tensors and run eager DDP?

'inputs' here refers to all the grad tensors that ddp wants to bcast/allreduce?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 1 above)
Comment (alanwaketan) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1236]: LazyTensor should be `non_overlapping_and_dense`. But there is a bug preventing us from doing that: https://github.com/pytorch/pytorch/blob/master/torch/csrc/lazy/core/tensor_impl.cpp#L78.

File: torch/csrc/distributed/c10d/reducer.cpp [hunk 4]
```diff
@@ -326,7 +326,7 @@ void Reducer::check_grad_layout(
         bucket_view.strides());
   }
   if (!gradient_as_bucket_view_) {
-    TORCH_INTERNAL_ASSERT(!grad.is_alias_of(bucket_view));
+    // TORCH_INTERNAL_ASSERT(!grad.is_alias_of(bucket_view));
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/reducer.cpp:329]: i think iiuc @bdhirsh was suggesting that in order to pass checks like this, lazy tensor needs to be wrapped in the functionalization infra, which would wrap our impl in an dummy functional impl that answers yes to is_alias_of whenever the underlying storage is the same.

Do you have an idea of how to get ddp working without doing that?

File: torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp (hunk 3 above)
Comment (alanwaketan) [reply torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:1342]: Yes! Even though I think ideally we need to also move the grads back to lazy for the optimizer, yet I don't get any complains within the optimizer. (I'm double checking.)

File: torch/nn/parallel/distributed.py [hunk 5]
```diff
@@ -925,8 +925,15 @@ def forward(self, *inputs, **kwargs):
                 self._check_global_requires_backward_grad_sync(is_joined_rank=False)
 
             if self.device_ids:
+                isLazy = inputs[0].device.type == 'lazy'
```
Comment (wconstab) [inline torch/nn/parallel/distributed.py:928]: i'm confused what this part is for. 

1) the model/weights/grads from demo script should already be lazy tensors right?  - so is this part trying to move tensors to the correct device index for some reason?
2) aren't you just going to convert back to cuda inside c++?

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 4 above)
Comment (alanwaketan) [reply torch/csrc/distributed/c10d/reducer.cpp:329]: I don't think I have any. So for the real solution, I probably need functionalization.

File: torch/nn/parallel/distributed.py (hunk 5 above)
Comment (alanwaketan) [reply torch/nn/parallel/distributed.py:928]: `self.to_kwargs` moves things back to cuda devices... Annoying...

File: torch/csrc/distributed/c10d/reducer.cpp (hunk 4 above)
Comment (wconstab) [reply torch/csrc/distributed/c10d/reducer.cpp:329]: ok. is your proposal to use the fallback way until we get functionalization, or you think there is another way to 'hack' around it and still enable overlapped comm+compute sooner?

Comment (alanwaketan) [reply torch/csrc/distributed/c10d/reducer.cpp:329]: In short, I don't have answers for you now. I need more tests to validate the current implementation as right now I can't tell what's missing in the context of the provided toy example. Things seem working.

**Review by wconstab — APPROVED** [review]
this looks good to me, nice work.  might not hurt to check it in actually, but up to you

File: torch/nn/parallel/distributed.py (hunk 5 above)
Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:928]: I'm trying to think of a better way to do this, as it is not ideal that lazy tensor details are explicitly encoded in DDP's implementation. 

As a first step, can we define a handler for this logic somewhere in lazy tensor codebase, and import and run this handler in DDP? That will at least abstract out the implementation a bit.

Comment (rohan-varma) [inline torch/nn/parallel/distributed.py:928]: nit: snake_case `is_lazy`?

**Review by rohan-varma — APPROVED** [review]

File: torch/nn/parallel/distributed.py (hunk 5 above)
Comment (alanwaketan) [reply torch/nn/parallel/distributed.py:928]: Yea, this is definitely not ideal. Maybe I can make `to_kwargs` more robust to identify lazy devices.

**Pull request merged**